site/content/src/main/scala/shapeless/AutoTypeClassExercises.scala

```diff
@@ -0,0 +1,171 @@
+package shapelessex
+
+import org.scalatest._
+import shapeless._
+
+trait Monoid[T] {
+  def zero: T
+  def append(a: T, b: T): T
+}
+
+object Monoid extends ProductTypeClassCompanion[Monoid] {
+  def mzero[T](implicit mt: Monoid[T]) = mt.zero
+
+  implicit def booleanMonoid: Monoid[Boolean] = new Monoid[Boolean] {
+    def zero = false
+    def append(a: Boolean, b: Boolean) = a || b
+  }
+
+  implicit def intMonoid: Monoid[Int] = new Monoid[Int] {
+    def zero = 0
+    def append(a: Int, b: Int) = a + b
+  }
+
+  implicit def doubleMonoid: Monoid[Double] = new Monoid[Double] {
+    def zero = 0.0
+    def append(a: Double, b: Double) = a + b
+  }
+
+  implicit def stringMonoid: Monoid[String] = new Monoid[String] {
+    def zero = ""
+    def append(a: String, b: String) = a + b
+  }
+
+  object typeClass extends ProductTypeClass[Monoid] {
+    def emptyProduct = new Monoid[HNil] {
+      def zero = HNil
+      def append(a: HNil, b: HNil) = HNil
+    }
+
+    def product[F, T <: HList](mh: Monoid[F], mt: Monoid[T]) = new Monoid[F :: T] {
+      def zero = mh.zero :: mt.zero
+      def append(a: F :: T, b: F :: T) = mh.append(a.head, b.head) :: mt.append(a.tail, b.tail)
+    }
+
+    def project[F, G](instance: ⇒ Monoid[G], to: F ⇒ G, from: G ⇒ F) = new Monoid[F] {
+      def zero = from(instance.zero)
+      def append(a: F, b: F) = from(instance.append(to(a), to(b)))
+    }
+  }
+}
+
+trait MonoidSyntax[T] {
+  def |+|(b: T): T
+}
+
+object MonoidSyntax {
+  implicit def monoidSyntax[T](a: T)(implicit mt: Monoid[T]): MonoidSyntax[T] = new MonoidSyntax[T] {
+    def |+|(b: T) = mt.append(a, b)
+  }
+}
+
+/** == Automatic type class instance derivation ==
+  *
+  * Based on and extending `Generic` and `LabelledGeneric`, Lars Hupel ([[https://twitter.com/larsr_h @larsr_h]]) has contributed the `TypeClass`
+  * family of type classes, which provide automatic type class derivation facilities roughly equivalent to those available
+  * with GHC as described in [[http://dreixel.net/research/pdf/gdmh.pdf "A Generic Deriving Mechanism for Haskell"]].  There is a description of an
+  * earlier iteration of the Scala mechanism [[http://typelevel.org/blog/2013/06/24/deriving-instances-1.html here]], and examples of its use deriving `Show` and `Monoid`
+  * instances [[https://github.com/milessabin/shapeless/blob/master/examples/src/main/scala/shapeless/examples/shows.scala here]]
```

Comment:
We can make these links independent of Shapeless' `master` branch because those files may be deleted. This can be done visiting the URL and pressing `y`, it will redirect you to a URL with the hash of the blob you are looking at like https://github.com/milessabin/shapeless/blob/f1d763732ac56fed1a811bc0f301d96f3c81bb7b/examples/src/main/scala/shapeless/examples/shows.scala

This way the links will not break and always point at what we were looking at when writing this.